From 8565402b07cb83b865c0686bd4aa2512bf0b9d19 Mon Sep 17 00:00:00 2001
From: Tim Hartley <timothy.d.hartley@intel.com>
Date: Wed, 11 Sep 2013 15:56:14 -0700
Subject: Dalvik: PCG GL fix for bad VR load in presence of registerize MIRs

BZ: 137896

Prior to this patch, the PCG GL only tracked references to use 0 in
kMirOpRegisterize MIRs. This led to a situation where higher VRs
aliased with VR0, leading to an incorrect load for a VR.

The symptom was an excessive number of iterations of the inner loop on
Smartbench Pi, caused by the loop bound being much higher than it
ought to be.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: Iad607f74fc66e9522d489a84eb76f263830a627c
Orig-MCG-Change-Id: Ie641b1b4167ee66c50aba74ade7681a020b22e3e
Signed-off-by: Tim Hartley <timothy.d.hartley@intel.com>
Reviewed-on: http://android.intel.com:8080/131494
Reviewed-by: Banerji, Udayan <udayan.banerji@intel.com>
Reviewed-by: Dittert, Eric <eric.dittert@intel.com>
Reviewed-by: Lupusoru, Razvan A <razvan.a.lupusoru@intel.com>
Reviewed-by: Kreitzer, David L <david.l.kreitzer@intel.com>
Reviewed-by: Beyler, Jean Christophe <jean.christophe.beyler@intel.com>
Reviewed-by: Chen, Dong-Yuan <dong-yuan.chen@intel.com>
Reviewed-by: lab_aqa <lab_aqa@intel.com>
Reviewed-by: Popov, Ivan G <ivan.g.popov@intel.com>
Tested-by: Chen, Dong-Yuan <dong-yuan.chen@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/codegen/x86/pcg/Analysis.cpp |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/vm/compiler/codegen/x86/pcg/Analysis.cpp b/vm/compiler/codegen/x86/pcg/Analysis.cpp
index 8e28fe8..4c2b489 100644
--- a/vm/compiler/codegen/x86/pcg/Analysis.cpp
+++ b/vm/compiler/codegen/x86/pcg/Analysis.cpp
@@ -1472,7 +1472,10 @@ bool dvmCompilerPcgNewRegisterizeVRAnalysis (CompilationUnitPCG *cUnit)
                     break;
 
                 case kMirOpRegisterize:
-                    pcgRef (cUnit, ssaRep->uses[0], Any);
+                    for (int i = 0; i < mir->ssaRep->numUses; i++)
+                    {
+                        pcgRef (cUnit, ssaRep->uses[i], Any);
+                    }
                     break;
 
                 case kMirOpPhi:
-- 
1.7.4.1

